### PR TITLE
cls_rgw: use a template to read input op

### PR DESCRIPTION
**Pull request by theanalyst**
While writing up a new cls method for LC to get a key, I came across a repeated pattern we use everywhere. This is an attempt to reduce that.

Signed-off-by: Abhishek Lekshmanan <abhishek@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by theanalyst** [comment]
Think I got most of them, only compile tested so far.

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -809,17 +809,27 @@ static int read_key_entry(cls_method_context_t hctx, cls_rgw_obj_key& key,
   return 0;
 }
 
-int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
-{
-  // decode request
-  rgw_cls_obj_complete_op op;
-  auto iter = in->cbegin();
+template <typename T>
+static int decode_input_op(bufferlist *in, T& op){
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc]: prefer bufferlist* (or buffer::list*) for new code if we can

Comment (theanalyst) [reply src/cls/rgw/cls_rgw.cc]: ack

**Comment by theanalyst** [comment]
actually this pattern is repeated in cls_rgw_client as well, I guess we could unify all the decode* methods under this?

**Comment by tchaikov** [comment]
@theanalyst ping?

**Pull request closed without merging**